experimental/calendar/src/date.rs

```diff
@@ -0,0 +1,123 @@
+// This file is part of ICU4X. For terms of use, please see the file
+// called LICENSE at the top level of the ICU4X source tree
+// (online at: https://github.com/unicode-org/icu4x/blob/main/LICENSE ).
+
+use crate::{Calendar, DateDuration, DurationUnit, Iso};
+use std::fmt;
+
+pub trait AsCalendar {
+    type Calendar: Calendar;
+    fn as_calendar(&self) -> &Self::Calendar;
+}
+
+impl<C: Calendar> AsCalendar for C {
+    type Calendar = C;
+    fn as_calendar(&self) -> &Self {
+        self
+    }
+}
+
+pub struct Date<A: AsCalendar> {
+    inner: <A::Calendar as Calendar>::DateInner,
+    calendar: A,
+}
+
+impl<A: AsCalendar> Date<A> {
+    /// Construct a date from an ISO date and some calendar representation
+    pub fn new_from_iso(iso: Date<Iso>, calendar: A) -> Self {
+        let inner = calendar.as_calendar().date_from_iso(iso);
+        Date { inner, calendar }
+    }
+
+    pub fn to_iso(&self) -> Date<Iso> {
+        self.calendar.as_calendar().date_to_iso(self.inner())
+    }
+
+    /// The number of months in the year of this date
+    pub fn months_in_year(&self) -> u8 {
+        self.calendar.as_calendar().months_in_year(&self.inner)
+    }
+
+    /// The number of days in the year of this date
+    pub fn days_in_year(&self) -> u32 {
+        self.calendar.as_calendar().days_in_year(&self.inner)
+    }
+
+    /// The number of days in the month of this date
+    pub fn days_in_month(&self) -> u8 {
+        self.calendar.as_calendar().days_in_month(&self.inner)
+    }
+
+    /// The day of the week for this date
+    ///
+    /// Monday is 1, Sunday is 7, according to ISO
+    pub fn day_of_week(&self) -> u8 {
+        self.calendar.as_calendar().day_of_week(&self.inner)
+    }
+
+    /// Add a `duration` to this date, mutating it
```

Comment:
Suggestion: Add both a `&mut self` and a `self` version of these types of functions, like I did in FixedDecimal with `negate(&mut self)` and `negated(self)`.